Add 32f_s32f_add_32f kernel

ghostop14 opened the pull request:
This kernel adds in vector + scalar functionality

ghostop14 commented:
If you wouldn't mind merging this one first, I'll incorporate it into #364 before merging that one.  Thanks!

argilo commented:
CI fails with this error:
 ```
libvolk.so.2.2.1: undefined reference to `volk_32f_s32f_add_32f_a_orc_impl'
```

argilo commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,295 @@
+ *   // Normalize by the smallest delta (0.2 in this example)
+ *   float scale = 5.0f;
+ *
+ *   volk_32f_s32f_add_32f(out, increasing, scale, N);
```
This example is no longer normalizing / scaling the input, so this example should be updated.

argilo commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,295 @@
+
+  for(number = 0; number < quarterPoints; number++){
+    aVal = vld1q_f32(inputPtr); // Load into NEON regs
+    cVal = vaddq_n_f32 (aVal, scalar); // Do the multiply
```
```suggestion
    cVal = vaddq_n_f32 (aVal, scalar); // Do the add
```

argilo commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,295 @@
+  volk_32f_s32f_add_32f_a_orc_impl(cVector, aVector, scalar, num_points);
+}
+
+#endif /* LV_HAVE_GENERIC */
```
This typo was copied over from the scalar multiply kernel:
```suggestion
#endif /* LV_HAVE_ORC */
```

argilo commented:
The reason for the CI failure is that `kernels/volk/asm/orc/volk_32f_s32f_multiply_32f_a_orc_impl.orc` was not copied over and adapted for addition. Either that should be done, or the `#ifdef LV_HAVE_ORC` section should be removed.

argilo commented:
I notice there are two separate include guards: `INCLUDED_volk_32f_s32f_add_32f_u_H` and `INCLUDED_volk_32f_s32f_add_32f_a_H`. The NEON kernel is in the latter section despite being an unaligned kernel. (This happened because it's the same in the multiply const kernel.)

michaelld commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,297 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2012, 2014, 2020 Free Software Foundation, Inc.
```
just the current year for a new file

michaelld commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,297 @@
+ *   float* increasing = (float*)volk_malloc(sizeof(float)*N, alignment);
+ *   float* out = (float*)volk_malloc(sizeof(float)*N, alignment);
+ *
+ *
```
remove the extra line here

michaelld commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,297 @@
+#include <arm_neon.h>
+
+static inline void
+volk_32f_s32f_add_32f_u_neon(float* cVector, const float* aVector,
```
is there an aligned version for neon?

michaelld commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,297 @@
+
+  for(number = 0; number < quarterPoints; number++){
+    aVal = vld1q_f32(inputPtr); // Load into NEON regs
+    cVal = vaddq_f32 (aVal, scalarvec); // Do the add
```
extra space between "f32" and "(" .. not a big deal since we'll do a clang-format soon-ish ... but still like to get it fixed where reasonable

michaelld reviewed (changes requested):
A few minor changes. looks pretty good right already though!

ghostop14 commented:
@michaelld all cleaned up.  On the a_neon question, this set was basically a copy-and-change version of the existing multiply block to get the add functionality.  Multiply didn't have one to use as a foundation either.

michaelld commented:
> @michaelld all cleaned up. On the a_neon question, this set was basically a copy-and-change version of the existing multiply block to get the add functionality. Multiply didn't have one to use as a foundation either.

OK; that makes sense about the neon. 'u' is a superset of 'a' anyway, so having 'u' working is the way to go if there is just 1 implementation.

michaelld reviewed (approved):
LGTM

michaelld commented:
@jdemel Let's get your review here to get this PR finalized, so that @ghostop14 can then integrate it into #364 & we can move on that.

ghostop14 commented:
Argilo and I actually removed the requirement in 364.  The add 1e-20 wasn't necessary.  So they can actually be approved separately (no longer coupled).

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,296 @@
+    outputPtr++;
+  }
+}
+#endif /* LV_HAVE_GENERIC */
```
I suggest, we remove the "aligned" version.

As far as I recall, we don't use separate include guards for aligned and unaligned kernels. Every kernel just gets its `#ifdef LV_HAVE_*`. That's probably an artifact from former times.

A few more things regarding the generic kernel.
1. Please move the generic kernel to the top. In most cases it serves as further documentation for others who want to understand the kernel in depth.
2. Let's just have one `generic` kernel and not aligned and unaligned.
3. Please use the generic kernel to take care of all the remainders in the other kernels. Mostly, because this avoids copypasta. I know that's not much in this case but I want to stick to good code practice and all code should serve as an example. Especially whenever someone's working on it.

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,296 @@
+  number = eighthPoints * 8;
+  for(;number < num_points; number++){
+    *cPtr++ = (*aPtr++) + scalar;
+  }
```
Just use the generic kernel here.

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,296 @@
+  number = quarterPoints * 4;
+  for(;number < num_points; number++){
+    *cPtr++ = (*aPtr++) + scalar;
+  }
```
Just use the generic kernel here

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,296 @@
+#define INCLUDED_volk_32f_s32f_add_32f_a_H
+
+#include <inttypes.h>
+#include <stdio.h>
```
I suggest to remove the extra include guards for aligned/unaligned kernels.
In this case, we can avoid these double includes.

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,296 @@
+  }
+  for(number = quarterPoints * 4; number < num_points; number++){
+    *outputPtr++ = (*inputPtr++) + scalar;
+  }
```
Please use the generic kernel here.

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,296 @@
+  }
+}
+
+#endif /* LV_HAVE_GENERIC */
```
This generic kernel should be sufficient. No need for an aligned generic version.

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,296 @@
+  number = eighthPoints * 8;
+  for(;number < num_points; number++){
+    *cPtr++ = (*aPtr++) + scalar;
+  }
```
Please use the generic kernel.

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,296 @@
+  number = quarterPoints * 4;
+  for(;number < num_points; number++){
+    *cPtr++ = (*aPtr++) + scalar;
+  }
```
Please use the generic kernel

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,296 @@
+#define INCLUDED_volk_32f_s32f_add_32f_u_H
+
+#include <inttypes.h>
+#include <stdio.h>
```
Please consolidate these include guards such that we only have one for all kernel versions.

jdemel reviewed (changes requested):
Overall this PR looks good to me. Just a few minor things. Thanks for your patience and your work.

ghostop14 commented:
@jdemel, just pushed updates.  I think I checked off all your comments.  Let me know if I missed anything?

ghostop14 committed:
Add 32f_s32f_add_32f kernel

This kernel adds in vector + scalar functionality

ghostop14 commented:
Rebased, fixed merge issues, clang-formatted my modified files, and re-pushed.  Doesn't look like CI re-ran after force push but the conflicts are resolved in the repo.

jdemel commented:
There seems to be an issue with GitHub. These are the current stats: `+32,918 −27,686`. I'd be really hesitant to just hit the merge button here. I assume git will be more helpful here.

michaelld commented:
Looks like the rebase didn't fully succeed ... please give it a go again when you get a chance!

ghostop14 committed:
Merge branch 'master' into addkernel

jdemel commented on kernels/volk/volk_32f_s32f_add_32f.h:
```diff
@@ -0,0 +1,275 @@
+#endif /* INCLUDED_volk_32f_s32f_add_32f_u_H */
+
+
+#ifndef INCLUDED_volk_32f_s32f_add_32f_a_H
+#define INCLUDED_volk_32f_s32f_add_32f_a_H
```
The include guards should be consolidated into one. For me, this is mainly a matter of consistency and it would be great to have it in here.

jdemel reviewed (changes requested):
I have one, hopefully minor comment and then this PR is good to go.

ghostop14 committed:
Merge remote-tracking branch 'upstream/master' into addkernel

ghostop14 committed:
This kernel adds in vector + scalar functionality

ghostop14 commented:
@jdemel take a look at this version.  I also rebased on master before pushing for this one.  I think the guards were okay for aligned versus unaligned with the exception of the ORC kernels.  Both aligned and unaligned ORC functions were in the aligned section, so I fixed that.  Not sure if that was exactly what you saw or not.  FYI, this kernel was templated off of the scalar multiply kernel, so the same guard issue is probably in some of the other scalar kernels too.  (Future cleanup task :) )

jdemel commented:
@ghostop14 Thanks for the rebase. Besides, I think there's a reason for ORC kernels to have aligned/unaligned include guards. At least, CI tells us that the kernel is broken now. It won't even compile on ARM. I wonder why though. Anyways. I assume, this is something we need to investigate at a different time. 
I'm OK with this PR. I guess you can revert your last commit where you moved the ORC kernel function.

ghostop14 commented:
@jdemel done.  We'll just make sure it builds in CI (famous last words... built fine on my machine)

michaelld commented:
CI error that I think needs to be fixed up herein:
```
In file included from /home/travis/build/gnuradio/volk/build/lib/volk_machine_neon.c:81:0:
/home/travis/build/gnuradio/volk/kernels/volk/volk_32f_s32f_add_32f.h: In function ‘volk_32f_s32f_add_32f_u_neon’:
/home/travis/build/gnuradio/volk/kernels/volk/volk_32f_s32f_add_32f.h:184:35: error: ‘cPtr’ undeclared (first use in this function)
     volk_32f_s32f_add_32f_generic(cPtr, aPtr, scalar, num_points - number);
                                   ^
/home/travis/build/gnuradio/volk/kernels/volk/volk_32f_s32f_add_32f.h:184:35: note: each undeclared identifier is reported only once for each function it appears in
/home/travis/build/gnuradio/volk/kernels/volk/volk_32f_s32f_add_32f.h:184:41: error: ‘aPtr’ undeclared (first use in this function)
     volk_32f_s32f_add_32f_generic(cPtr, aPtr, scalar, num_points - number);
                                         ^
make[2]: *** [lib/CMakeFiles/volk_obj.dir/volk_machine_neon.c.o] Error 1
make[1]: *** [lib/CMakeFiles/volk_obj.dir/all] Error 2
make: *** [all] Error 2
```

ghostop14 commented:
@michaelld looks good now.

jdemel reviewed (approved):
LGTM. Thanks for working on it!

michaelld reviewed (approved):
LGTM

michaelld commented:
Relevant CI passes. 2 approvals. Nice addition! Merging
